src/compiler/checker.ts

```diff
@@ -7845,6 +7851,16 @@ namespace ts {
                 if (isBindingPattern(declaration.name)) {
                     return getTypeFromBindingPattern(<BindingPattern>declaration.name, /*includePatternInType*/ true);
                 }
+                if (isBindingPattern(declaration.parent)) {
+                    const parentDeclaration = declaration.parent.parent;
+                    const name = declaration.propertyName || declaration.name;
+                    if (isVariableLike(parentDeclaration) &&
+                        parentDeclaration.type &&
+                        (name.kind === SyntaxKind.Identifier || name.kind == SyntaxKind.StringLiteral)) {

```

Comment:
This JSON-like syntax is allowed, which is the StringLiteral case, I believe:

```ts
function f( { "prop": local = v => +v }: Foo) {
}
```

The code you give is a ComputedPropertyName, I believe. I'll test all four.

Also, I used `==` instead of `===` by mistake. I'm not sure how the linter missed that.